### PR TITLE
Resolving subqueries columns

### PR DESCRIPTION
**Pull request by collerek**
Final piece of work in regard of resolving aliases and columns coming from sub-queries to the actual database columns.
Now all aliases should resolve and only real columns should be included in columns and columns_dict.

Following the updated readme note the tables in join:
```python
from sql_metadata import Parser

parser = Parser(
"""
SELECT COUNT(1) FROM
(SELECT std.task_id FROM some_task_detail std WHERE std.STATUS = 1) a
JOIN (SELECT st.task_id FROM some_task st WHERE task_type_id = 80) b
ON a.task_id = b.task_id;
"""
)

# get sub-queries dictionary
parser.subqueries
# {"a": "SELECT std.task_id FROM some_task_detail std WHERE std.STATUS = 1",
#  "b": "SELECT st.task_id FROM some_task st WHERE task_type_id = 80"}


# get names/ aliases of sub-queries / derived tables
parser.subqueries_names
# ["a", "b"]

# note that columns coming from sub-queries are resolved to real columns
parser.columns
#["some_task_detail.task_id", "some_task_detail.STATUS", "some_task.task_id", 
# "task_type_id"]

# same applies for columns_dict, note the join columns are resolved
parser.columns_dict
#{'join': ['some_task_detail.task_id', 'some_task.task_id'],
# 'select': ['some_task_detail.task_id', 'some_task.task_id'],
# 'where': ['some_task_detail.STATUS', 'task_type_id']}

```

Let me know what you think :)

### CODE REVIEW

File: sql_metadata/token.py [hunk 1]
```diff
@@ -169,10 +169,26 @@ def is_alias_without_as(self) -> bool:
...
+        Returns if current token is a definition of an alias.
+        Note that aliases can also be used in other queries and be a part
+        of other nested columns with aliases
+        """
+        return (
+            self.is_alias_without_as
+            or self.previous_token.normalized == "AS"
+            or self.is_in_with_columns
+        )
+
```
Comment (macbre) [inline sql_metadata/token.py:191]: 👌🏻

File: sql_metadata/token.py [hunk 2]
```diff
@@ -199,6 +215,18 @@ def is_in_with_columns(self) -> bool:
...
+        """
+        Determines if * encountered in query is a wildcard like select <*> from aa
+        or is that an operator like Select aa <*> bb as cc from dd
+        """
+        return self.normalized == "*" and (
+            self.previous_token.value in [",", ".", "SELECT"]
+            or (self.previous_token.value == "(")
+            and self.next_token.value == ")"
+        )
+
```
Comment (macbre) [inline sql_metadata/token.py:229]: 👌🏻

File: sql_metadata/parser.py [hunk 3]
```diff
@@ -714,6 +729,37 @@ def _resolve_alias_to_column(self, alias_token: SQLToken) -> str:
...
+        """
+        Resolves subquery reference to the actual column in the subquery
+        """
+        parts = subquery_alias.split(".")
+        if len(parts) != 2 or parts[0] not in self.subqueries_names:
+            return [subquery_alias]
+        sub_qry, column_name = parts[0], parts[-1]
+        sub_qry_definition = self.subqueries.get(sub_qry)
+        subparser = self._subqueries_parsers.setdefault(
+            sub_qry, Parser(sub_qry_definition)
```
Comment (macbre) [inline sql_metadata/parser.py:742]: Can we use here `sub_query` instead of a short name (`sub_qry`) that only save us two letters?

Comment (collerek) [reply sql_metadata/parser.py:742]: Sure, will update

Comment (collerek) [reply sql_metadata/parser.py]: Done

**Comment by collerek** [comment]
I realized that I forgot about with queries -> so that's also something that needs to be developed later (resolving with queries).

File: README.md [hunk 4]
```diff
@@ -7,14 +7,22 @@
 [![Downloads](https://pepy.tech/badge/sql-metadata/month)](https://pepy.tech/project/sql-metadata)
 
 Uses tokenized query returned by [`python-sqlparse`](https://github.com/andialbrecht/sqlparse) and generates query metadata.
-**Extracts column names and tables** used by the query. Provides a helper for **normalization of SQL queries** and **tables aliases resolving**.
+
+**Extracts column names and tables** used by the query. 
+Automatically conduct **column alias resolution**, **sub queries aliases resolution** as well as **tables aliases resolving**.
+
+Provides also a helper for **normalization of SQL queries**.
```
Comment (macbre) [inline README.md:14]: 👌🏻

**Comment by collerek** [comment]
@macbre  - now after coverage add the test pipeline fails with lack of GITHUB_TOKEN

**Comment by collerek** [comment]
@macbre - Ok I fixed the coverall runs, do you need me to do anything more for this PR?

File: sql_metadata/utils.py [hunk 5]
```diff
@@ -13,5 +13,22 @@ def append(self, item: Any) -> None:
         if item not in self:
             super().append(item)
 
+    def extend(self, items: Any) -> None:
```
Comment (macbre) [inline sql_metadata/utils.py:16]: Let's make it explicit that `items` is expected to be [the sequence](https://docs.python.org/3/library/stdtypes.html#typesseq).

```suggestion
    def extend(self, items: Sequence[Any]) -> None:
```

File: sql_metadata/token.py [hunk 6]
```diff
@@ -217,7 +250,7 @@ def table_prefixed_column(self, table_aliases: Dict) -> str:
         value = self.left_expanded
         if "." in value:
             parts = value.split(".")
-            if len(parts) > 3:  # pragma: no cover
+            if len(parts) > 2:  # pragma: no cover
```
Comment (macbre) [inline sql_metadata/token.py:253]: Did we stop supporting `database.table.column` notation?

**Review by macbre — APPROVED** [review]

File: sql_metadata/token.py (hunk 6 above)
Comment (collerek) [reply sql_metadata/token.py:253]: When i checked some backends it appeared it's not possible. You can have 3 part table names but not column names. If you have some backend that supports it we can revert.

Comment (collerek) [reply sql_metadata/token.py:253]: But then we don't have any test for it :)

Comment (macbre) [reply sql_metadata/token.py:253]: Makes sense, thanks for clarifying 🙂

**Pull request merged**